frontend/src/components/common/content/ContentBlockRenderer.ts

```diff
@@ -0,0 +1,64 @@
+import React from "react";
+import { ContentType } from "../../../APIClients/types/LessonClientTypes";
+import { ContentBlockState } from "../../../types/ContentBlockTypes";
+
+// Many thanks to https://github.com/Microsoft/TypeScript/issues/30581#issuecomment-1080979994.
+
+type ValueOf<T> = T[keyof T];
+
+type ArgMap<D extends ContentType> = {
+  [K in D]: Extract<ContentBlockState, { ["type"]: { ["clientType"]: K } }>;
+};
+
+class ContentBlockRenderer<
```

Comment:
It might be good to have a comment talking about why this is necessary so it's less cryptic (e.g. so we can have a table instead of a switch case or "Implements a table that takes in a component block type and renders a coressponding React component")